### PR TITLE
Show number of subscribers when editing and creating segments [MAILPOET-3219]

### PR DESCRIPTION
**Pull request by lysyjan**
[MAILPOET-3219]

[MAILPOET-3219]: https://mailpoet.atlassian.net/browse/MAILPOET-3219

### CODE REVIEW

File: assets/js/src/segments/subscribers_counter.tsx [hunk 1]
```diff
@@ -0,0 +1,55 @@
+import React from 'react';
+import MailPoet from 'mailpoet';
+
+type SubscriberCount = {
```
Comment (pavel-mailpoet) [inline assets/js/src/segments/subscribers_counter.tsx:4]: In the calculator, you were using a different syntax for the type. You used an interface and number | undefined. I think we should start unifying it. So instead of using {} here, I would recommend using `interface`. and instead of  using number | undefined I would use `?` syntax

File: assets/js/src/segments/subscribers_calculator.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  wooCommerceActions: string[] = ['purchasedCategory', 'purchasedProduct'];
+
+  // Names of keys from interface FormItem
+  allowedItemKeys: string[] = ['wordpressRole', 'segmentType', 'action', 'newsletter_id', 'category_id', 'product_id'];
+
+  lastFormItem: FormItem;
+
+  lastResult: Result;
+
+  validateInputData(formItem: FormItem): boolean {
```
Comment (pavel-mailpoet) [inline assets/js/src/segments/subscribers_calculator.ts:36]: I am not sure about having this here. The idea, in the beginning, was to keep filter specific logic in the `dynamic_segments_filters` directory. The reason is to keep it in one place and when we are going to add new segments we will have only a couple of files to modify. If we keep adding logic like this, adding new segments will be more complicated. 

The same applies to the actions above.

File: assets/js/src/segments/subscribers_calculator.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
+import MailPoet from 'mailpoet';
+import _ from 'lodash';
+
+export interface FormItem {
```
Comment (pavel-mailpoet) [inline assets/js/src/segments/subscribers_calculator.ts:4]: Is it possible to have a better type? The item shouldn't have category_id and `wordpressRole` at the same time, right?

Also, can we rename the items so that we don't combine underscore and camelcase?

File: lib/Segments/SegmentSubscribersRepository.php [hunk 4]
```diff
@@ -85,6 +86,18 @@ public function getSubscribersCountBySegmentIds(array $segmentIds, string $statu
     return (int)$result;
   }
 
+  public function getDynamicSubscribersCount(DynamicSegmentFilterData $data): int {
```
Comment (pavel-mailpoet) [inline lib/Segments/SegmentSubscribersRepository.php:89]: We already have a function for calculating subscribers in dynamic segments. When we display segments we almost always display a count of subscribers next to them. Can't we reuse that? Why do we need a new one? 

This function calculates all subscribers and doesn't filter out deleted, doesn't check status. Is that intentional?

**Review by pavel-mailpoet — CHANGES_REQUESTED** [review]

File: assets/js/src/segments/dynamic_segments_form.jsx [hunk 5]
```diff
@@ -102,12 +115,48 @@ class DynamicSegmentForm extends React.Component {
     }
   }
 
+  getCount() {
+    const { item } = this.state;
+    switch (item.segmentType) {
```
[the remark this answers is not in the document]
Comment (lysyjan) [reply assets/js/src/segments/dynamic_segments_form.jsx:120]: Of course, there can be `if`. My first idea was a little bit different than the final solution.

[the remark this answers is not in the document]
Comment (pavel-mailpoet) [reply assets/js/src/segments/dynamic_segments_form.jsx:120]: I mean, we don't need anything here, right? The calculator does check the value anyway.

File: assets/js/src/segments/subscribers_calculator.ts (hunk 2 above)
Comment (lysyjan) [reply assets/js/src/segments/subscribers_calculator.ts:36]: I understand, but how would you solve that we don't want to display the message about calculating and don't call API without all data?

Comment (lysyjan) [reply assets/js/src/segments/subscribers_calculator.ts:36]: I get it. There is a place for some refactoring maybe?

Comment (pavel-mailpoet) [reply assets/js/src/segments/subscribers_calculator.ts:36]: I would still have the validation, but move it in the filters. Each filter could validate its own data and return false/true if the data is valid. And having the main validation function that would call dynamically the filter validation based on the type.

File: assets/js/src/segments/subscribers_calculator.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  errors: string[];
+}
+
+interface ApiVersionWindow extends Window {
+  mailpoet_api_version: string;
+}
+
+declare let window: ApiVersionWindow;
+
+class SubscribersCalculator {
```
[the remark this answers is not in the document]
Comment (lysyjan) [reply assets/js/src/segments/subscribers_calculator.ts:24]: I used class because I wanted to store a previous state. I don't want to call API when the name of the description is changed. I would be glad when you help me with a better solution.

[the remark this answers is not in the document]
Comment (pavel-mailpoet) [reply assets/js/src/segments/subscribers_calculator.ts:24]: you can do that even with a function:
```
let previous = undefined;

function SubscribersCalculator() {
  if (previous) return previous;
  ... 
}
```

File: assets/js/src/segments/subscribers_calculator.ts (hunk 3 above)
Comment (lysyjan) [reply assets/js/src/segments/subscribers_calculator.ts:4]: > The item shouldn't have category_id and wordpressRole at the same time, right?

How can I do it? Could you help me, please?

> Also, can we rename the items so that we don't combine underscore and camelcase?

I used current input names from the form.

Comment (pavel-mailpoet) [reply assets/js/src/segments/subscribers_calculator.ts:4]: > I used current input names from the form.

I know and I think we should rename those input names to make them consistent.

Comment (pavel-mailpoet) [reply assets/js/src/segments/subscribers_calculator.ts:4]: Could we use something like this?
```
interface FormItem { 
  segmentType: string; 
} 

interface WordpressRoleFormItem extends FormItem { 
   wordpressRole?: string;
} 

interface WooCommerceFormItem extends FormItem { 
 ...
}

type AnyFormItem = WordpressRoleFormItem | WooCommerceFormItem

let lastFormItem: AnyFormItem;
```

File: lib/Segments/SegmentSubscribersRepository.php (hunk 4 above)
Comment (lysyjan) [reply lib/Segments/SegmentSubscribersRepository.php:89]: I have overlooked this question. I can't reuse the same method because the dynamic segment isn't created or updated in the DB. I know only the filter parameters. I will try to change it a little bit.

File: assets/js/src/ajax.js [hunk 7]
```diff
@@ -55,8 +59,8 @@ MailPoet.Ajax = {
     }
 
     // set default timeout
-    if (this.options.token === null) {
-      this.options.timeout = 0;
+    if (this.options.timeout === null) {
+      this.options.timeout = 2000;
```
Comment (pavel-mailpoet) [inline assets/js/src/ajax.js:63]: Is this change intentional? Do you really want to limit all requests to 2 seconds?

Comment (lysyjan) [reply assets/js/src/ajax.js:63]: This is definitely my mistake. I tested an error message and I didn't notice this change after that. Thank you

**Review by pavel-mailpoet — APPROVED** [review]

**Review by veljkho — CHANGES_REQUESTED** [review]
Everything's good except clicked links in a segment.

For example:

I have 2 subscribers who clicked on different links in an email.
Then I create a `Segment` -> `Select` action `clicked` and I see **2 subscribers** (which is correct). Then I choose `links` and always see **[2 subscribers](https://d.pr/i/DNxjda)**? Is that expected? asking because I expect to see **1 subscriber**, because only 1 has clicked on that specific link chosen in the dropdown. When I save it I see **[1 subscriber](https://d.pr/i/bopecb)** on Segments page, so its a bit confusing for users to see different values in editor and who is really counted.

**Review by pavel-mailpoet — APPROVED** [review]

**Comment by lysyjan** [comment]
@veljkho Nice catch 👍🏻

**Review by veljkho — DISMISSED** [review]

**Review by veljkho — APPROVED** [review]
